src/Agent/NewRelic/Agent/Core/Utilities/GuidGenerator.cs

```diff
@@ -63,12 +53,48 @@ private static string GenerateTraceId()
 
         private static string GetTraceIdFromCurrentActivity()
         {
-            if (Activity.Current != default && Activity.Current.IdFormat == ActivityIdFormat.W3C)
-            {
-                return Activity.Current.TraceId.ToString();
-            }
+            // because we ILRepack System.Diagnostics.DiagnosticSource, we have to look for the app's reference to it (if there is one)
+            // and use reflection to get the trace id from the current activity
+
+            // get list of loaded assemblies
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            // find System.Diagnostics.DiagnosticSource
+            var diagnosticSourceAssembly = Array.Find(assemblies, a => a.FullName.StartsWith("System.Diagnostics.DiagnosticSource"));
+            if (diagnosticSourceAssembly == null) // customer app didn't reference the assembly
+                return null;
+
+            // find the Activity class
+            var activityType = diagnosticSourceAssembly.GetType("System.Diagnostics.Activity");
+
+            var fieldReadAccessor = VisibilityBypasser.Instance.GenerateFieldReadAccessor<object>(activityType, "s_current");
+            if (fieldReadAccessor == null)
+                return null;
+
+            var current = fieldReadAccessor(null);
+            if (current == null)
+                return null;
+
+            // get the Value property
+            var valuePropertyAccessor = VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(current.GetType(), "Value");
+            if (valuePropertyAccessor == null)
+                return null;
+
+            var value = valuePropertyAccessor(current);
+            if (value == null)
+                return null;
+
+            // get IdFormat property
+            var idFormatGetter = VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(value.GetType(), "IdFormat");
+            var idFormat = idFormatGetter(value);
+            if (idFormat == null || Enum.GetName(idFormat.GetType(), idFormat) != "W3C") // make sure it's in W3C trace id format
+                return null;
 
-            return GenerateTraceId();
+            // get TraceId property
+            var traceIdGetter = VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(value.GetType(), "TraceId");
+            if (traceIdGetter == null)
+                return null;
+                
+            return traceIdGetter(value).ToString();
```

Comment:
This seems like a good start but we probably need to cache these visibility bypasser calls, otherwise we'll continue to pay the reflection and code generation price with every call.

Reply:
Added caching, local testing works well.